tools/autograd/templates/Functions.cpp

```diff
@@ -408,6 +408,38 @@ Tensor smooth_l1_loss_double_backward_grad_output(const Tensor & grad, const Ten
   return (r * grad).sum().view({1});
 }
 
+inline int64_t diag_size(int64_t height, int64_t width, int64_t diagonal) {
+  if (width > height) {
+    return diag_size(width, height, -diagonal);
+  }
+  // Assumes height >= width
+  auto longest_diag = width;
+  if (diagonal >= 0) {
+    return longest_diag - diagonal;
+  }
+  if (longest_diag < height + diagonal) {
+    return longest_diag;
+  }
+  return height + diagonal;
+}
+
+Tensor diag_backward(const Tensor & grad, IntList input_sizes, int64_t diagonal) {
+  auto ndimension = input_sizes.size();
+  TORCH_ASSERT(ndimension == 1 || ndimension == 2);
+
+  if (ndimension == 1 || input_sizes[0] == input_sizes[1]) {
+    return grad.diag(diagonal);
+  }
+
+  // Input was a matrix but was not square
```

Comment:
That's not true, because `torch.diag` works in two modes. It can either select out a diagonal from a tensor, or, if given a 1d input, it will create a zero-filled square matrix, with the given diagonal. This case doesn't seem to be handled in here

Reply:
If the input is 1-dimensional, or the input is a square matrix, then it's handled by the above case with `return grad.diag(diagonal)`. If it's neither of those it falls through to this comment which takes care of the 2d and not-square case.

Reply:
Right, my bad...